<a href="https://colab.research.google.com/github/VitalyGladyshev/gb_nlp/blob/main/HW_02.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ДЗ 02

In [2]:
import numpy as np
import pandas as pd

In [3]:
from google.colab import drive
drive.mount('/content/gdrive')

Mounted at /content/gdrive


In [4]:
path_nlp = "/content/gdrive/My Drive/Colab Notebooks/NLP/"

In [5]:
train_cor = pd.read_pickle(path_nlp + "train_cor.pkl")
test_cor = pd.read_pickle(path_nlp + "test_cor.pkl")

In [6]:
train_cor.tail(2)

id  ...                                   tweet_lemmatized
31960  31961  ...  {act, vandalise, sikh, wso, temple, condemn, c...
31961  31962  ...                                    {thank, follow}

[2 rows x 8 columns]

In [7]:
test_cor.tail(2)

id  ...                                   tweet_lemmatized
49157  49158  ...  {conference, right, happy, cultureofdevelopmen...
49158  49159  ...  {free, song, glad, newmusic, download, shoegaz...

[2 rows x 7 columns]

In [8]:
test_cor['label'] = -1

In [9]:
data = train_cor.copy()
data = data.append(test_cor, ignore_index=True)
data.tail(2)

id  ...                                   tweet_lemmatized
49157  49158  ...  {conference, right, happy, cultureofdevelopmen...
49158  49159  ...  {free, song, glad, newmusic, download, shoegaz...

[2 rows x 8 columns]

## Задание 1

Создайте мешок слов с помощью sklearn.feature_extraction.text.CountVectorizer.fit_transform(). Применим его к 'tweet_stemmed' и 'tweet_lemmatized' отдельно.
- Игнорируем слова, частота которых в документе строго превышает порог 0.9 с помощью max_df.
- Ограничим количество слов, попадающий в мешок, с помощью max_features = 1000.
- Исключим стоп-слова с помощью stop_words='english'.
- Отобразим Bag-of-Words модель как DataFrame. columns необходимо извлечь с помощью CountVectorizer.get_feature_names().


In [10]:
from sklearn.feature_extraction.text import CountVectorizer

In [12]:
count_vectorizer_stemmed = CountVectorizer(stop_words='english', max_df=0.9, max_features=1000)
count_vectorizer_lemmatized = CountVectorizer(stop_words='english', max_df=0.9, max_features=1000)

In [28]:
list(map(lambda x: " ".join(x), data['tweet_stemmed'].values[:10]))

['kid drag father selfish dysfunct run',
 'getthank thank offer caus use pdx van lyft disapoint wheelchair credit',
 'majesti bihday',
 'love model ur time take',
 'factsguid societi motiv',
 'pay get leav fare disput allshowandnogo chao talk big fan huge',
 'danni camp tomorrow',
 'school year next think revolutionschool girl exam imagin hate actorslif',
 'cav love allin cleveland land champion clevelandcavali',
 'gr welcom']

In [29]:
bag_of_words_stemmed = count_vectorizer_stemmed.fit_transform(list(map(lambda x: " ".join(x), data['tweet_stemmed'].values)))
bag_of_words_lemmatized = count_vectorizer_lemmatized.fit_transform(list(map(lambda x: " ".join(x), data['tweet_lemmatized'].values)))

In [30]:
bag_of_words_stemmed

<49159x1000 sparse matrix of type '<class 'numpy.int64'>'
	with 202636 stored elements in Compressed Sparse Row format>

In [31]:
bag_of_words_lemmatized

<49159x1000 sparse matrix of type '<class 'numpy.int64'>'
	with 191072 stored elements in Compressed Sparse Row format>

In [41]:
df_count_vect_stemmed = pd.DataFrame(bag_of_words_stemmed.toarray(), columns=count_vectorizer_stemmed.get_feature_names())
df_count_vect_stemmed.head(2)

abl  absolut  accept  account  act  ...  young  youtub  yr  yrs  yummi
0    0        0       0        0    0  ...      0       0   0    0      0
1    0        0       0        0    0  ...      0       0   0    0      0

[2 rows x 1000 columns]

In [42]:
df_count_vect_lemmatized = pd.DataFrame(bag_of_words_lemmatized.toarray(), columns=count_vectorizer_lemmatized.get_feature_names())
df_count_vect_lemmatized.head(2)

able  absolutely  accept  account  act  ...  young  youtube  yr  yrs  yummy
0     0           0       0        0    0  ...      0        0   0    0      0
1     0           0       0        0    0  ...      0        0   0    0      0

[2 rows x 1000 columns]

## Задание 2

Создайте мешок слов с помощью sklearn.feature_extraction.text.TfidfVectorizer.fit_transform(). Применим его к 'tweet_stemmed' и 'tweet_lemmatized' отдельно.
- Игнорируем слова, частота которых в документе строго превышает порог 0.9 с помощью max_df.
- Ограничим количество слов, попадающий в мешок, с помощью max_features = 1000.
- Исключим стоп-слова с помощью stop_words='english'.
- Отобразим Bag-of-Words модель как DataFrame. columns необходимо извлечь с помощью TfidfVectorizer.get_feature_names().


In [33]:
from sklearn.feature_extraction.text import TfidfVectorizer

In [34]:
tfidf_vectorizer_stemmed = TfidfVectorizer(stop_words='english', max_df=0.9, max_features=1000)
tfidf_vectorizer_lemmatized = TfidfVectorizer(stop_words='english', max_df=0.9, max_features=1000)

In [35]:
bag_of_words_tfidf_stemmed = tfidf_vectorizer_stemmed.fit_transform(list(map(lambda x: " ".join(x), data['tweet_stemmed'].values)))
bag_of_words_tfidf_lemmatized = tfidf_vectorizer_lemmatized.fit_transform(list(map(lambda x: " ".join(x), data['tweet_lemmatized'].values)))

In [36]:
bag_of_words_tfidf_stemmed

<49159x1000 sparse matrix of type '<class 'numpy.float64'>'
	with 202636 stored elements in Compressed Sparse Row format>

In [37]:
bag_of_words_tfidf_lemmatized

<49159x1000 sparse matrix of type '<class 'numpy.float64'>'
	with 191072 stored elements in Compressed Sparse Row format>

In [43]:
df_tfidf_vect_stemmed = pd.DataFrame(bag_of_words_tfidf_stemmed.toarray(), columns=tfidf_vectorizer_stemmed.get_feature_names())
df_tfidf_vect_stemmed.head(2)

abl  absolut  accept  account  act  ...  young  youtub   yr  yrs  yummi
0  0.0      0.0     0.0      0.0  0.0  ...    0.0     0.0  0.0  0.0    0.0
1  0.0      0.0     0.0      0.0  0.0  ...    0.0     0.0  0.0  0.0    0.0

[2 rows x 1000 columns]

In [44]:
df_tfidf_vect_stemmed = pd.DataFrame(bag_of_words_tfidf_lemmatized.toarray(), columns=tfidf_vectorizer_lemmatized.get_feature_names())
df_tfidf_vect_stemmed.head(2)

able  absolutely  accept  account  act  ...  young  youtube   yr  yrs  yummy
0   0.0         0.0     0.0      0.0  0.0  ...    0.0      0.0  0.0  0.0    0.0
1   0.0         0.0     0.0      0.0  0.0  ...    0.0      0.0  0.0  0.0    0.0

[2 rows x 1000 columns]

## Задание 3

Натренируем gensim.models.Word2Vec модель на наших данных.
- Тренировать будем на токенизированных твитах combine_df['tweet_token']
- Установим следующие параметры: size=200, window=5, min_count=2, sg = 1, hs = 0, negative = 10, workers= 32, seed = 34.
- Используем функцию train() с параметром total_examples равным длине combine_df['tweet_token'], количество epochs установим 20.


In [45]:
from gensim.models import Word2Vec

In [50]:
model_w2v = Word2Vec(size=200, window=5, min_count=2, sg = 1, hs = 0, negative = 10, workers= 32, seed = 34)

In [51]:
model_w2v.build_vocab(data['tweet_token'].values)

In [52]:
model_w2v.train(data['tweet_token'].values, total_examples=len(data['tweet_token'].values), epochs=20, report_delay=1)

(9033885, 11669180)

In [53]:
model_w2v.wv.most_similar(positive=["car"])

/usr/local/lib/python3.6/dist-packages/gensim/matutils.py:737: FutureWarning: Conversion of the second argument of issubdtype from `int` to `np.signedinteger` is deprecated. In future, it will be treated as `np.int64 == np.dtype(int).type`.
  if np.issubdtype(vec.dtype, np.int):


[('truck', 0.5263153314590454),
 ('provo', 0.5135538578033447),
 ('caffine', 0.49682706594467163),
 ('crash', 0.49490049481391907),
 ('seat', 0.493663489818573),
 ('newadventure', 0.48260825872421265),
 ('glacier', 0.47989988327026367),
 ('accident', 0.47872379422187805),
 ('hsp', 0.47774893045425415),
 ('apament', 0.474631130695343)]

## Задание 4

Давайте немного потестируем нашу модель Word2Vec и посмотрим, как она работает. Мы зададим слово positive = "dinner", и модель вытащит из корпуса наиболее похожие слова c помощью функции most_similar. То же самое попробуем со словом "trump".

In [54]:
model_w2v.wv.most_similar(positive=["dinner"])

/usr/local/lib/python3.6/dist-packages/gensim/matutils.py:737: FutureWarning: Conversion of the second argument of issubdtype from `int` to `np.signedinteger` is deprecated. In future, it will be treated as `np.int64 == np.dtype(int).type`.
  if np.issubdtype(vec.dtype, np.int):


[('spaghetti', 0.5896735191345215),
 ('cookout', 0.5851148962974548),
 ('bihdaydinner', 0.5799993872642517),
 ('tacotuesday', 0.5735299587249756),
 ('saturdate', 0.5697401762008667),
 ('lunchtime', 0.5667805671691895),
 ('bolognese', 0.5645062923431396),
 ('lastnight', 0.5589499473571777),
 ('lamb', 0.5585891604423523),
 ('iftar', 0.5579138398170471)]

In [55]:
model_w2v.wv.most_similar(positive=["trump"])

/usr/local/lib/python3.6/dist-packages/gensim/matutils.py:737: FutureWarning: Conversion of the second argument of issubdtype from `int` to `np.signedinteger` is deprecated. In future, it will be treated as `np.int64 == np.dtype(int).type`.
  if np.issubdtype(vec.dtype, np.int):


[('donald', 0.5525873899459839),
 ('dumptrump', 0.5486760139465332),
 ('conman', 0.5186482071876526),
 ('suppoer', 0.5182057619094849),
 ('potus', 0.5109132528305054),
 ('trumptrain', 0.5054063200950623),
 ('donaldtrump', 0.5050008296966553),
 ('cochairman', 0.5049580335617065),
 ('presidentelect', 0.500296413898468),
 ('paladino', 0.4999353587627411)]

## Задание 5

Из приведенных выше примеров мы видим, что наша модель word2vec хорошо справляется с поиском наиболее похожих слов для данного слова. Но как она это делает? Она изучила векторы для каждого уникального слова наших данных и использует косинусное сходство, чтобы найти наиболее похожие векторы (слова).
<br>Давайте проверим векторное представление любого слова из нашего корпуса, например "food".

In [56]:
model_w2v.wv.word_vec("food")

array([-6.11505330e-01, -7.13534892e-01, -5.10297157e-02,  7.61005938e-01,
       -1.00077820e+00,  5.00927009e-02, -1.21215510e+00,  1.73618659e-01,
       -2.33678058e-01,  2.01332748e-01, -3.61983508e-01,  3.04014325e-01,
       -1.93961218e-01,  5.02443433e-01, -3.70827228e-01, -1.10314772e-01,
       -3.23230863e-01, -3.05364072e-01,  4.38812673e-01, -2.34569773e-01,
        1.43866271e-01,  6.59991920e-01, -3.37127447e-01, -4.44977403e-01,
        1.99118242e-01,  1.08627312e-01, -6.68186620e-02,  5.45131445e-01,
        3.95526201e-01, -1.76000565e-01,  2.09798083e-01, -4.15841073e-01,
        2.62166560e-01, -5.41864753e-01, -2.80287206e-01,  2.84253418e-01,
        5.43373585e-01,  3.89108770e-02,  5.21777391e-01, -2.71133363e-01,
       -2.64734812e-02, -4.01863128e-01,  8.35564196e-01, -1.51192263e-01,
        5.66735029e-01, -1.35796562e-01,  3.89534652e-01,  1.20460801e-01,
        1.97285458e-01, -6.76709354e-01,  7.62677118e-02,  3.73187631e-01,
       -3.99832847e-03,  

## Задание 6

Поскольку наши данные содержат твиты, а не только слова, нам придется придумать способ использовать векторы слов из модели word2vec для создания векторного представления всего твита. Существует простое решение этой проблемы, мы можем просто взять среднее значение всех векторов слов, присутствующих в твите. Длина результирующего вектора будет одинаковой, то есть 200. Мы повторим тот же процесс для всех твитов в наших данных и получим их векторы. Теперь у нас есть 200 функций word2vec для наших данных.
<br>Необходимо создать вектор для каждого твита, взяв среднее значение векторов слов, присутствующих в твите. В цикле сделать:  vec += model_w2v[word].reshape((1, size)) и поделить финальный вектор на количество слов в твите. На выходе должен получиться wordvec_df.shape = (49159, 200).


In [75]:
from typing import List

In [105]:
def w2v_mean(data: pd.Series)->List:
    wordvec_df = []
    size = 200
    for sent in data.values:
        vec = 0
        cnt = 0
        for word in sent:
            try:
                vec += model_w2v[word].reshape((1, size))
                cnt +=1
            except KeyError as err:
                pass
                # print(f"Exception type {type(err)}: {err}")
        if cnt:
            vec /= cnt
        wordvec_df.append(vec)
    return wordvec_df

In [106]:
res = w2v_mean(data['tweet_token'])

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:9: DeprecationWarning: Call to deprecated `__getitem__` (Method will be removed in 4.0.0, use self.wv.__getitem__() instead).
  if __name__ == '__main__':


In [107]:
len(res)

49159